pulsar-transformations/src/main/java/com/datastax/oss/pulsar/functions/transforms/jstl/JstlFunctions.java

```diff
@@ -44,4 +55,55 @@ public static String concat(Object first, Object second) {
   public static Object coalesce(Object value, Object valueIfNull) {
     return value == null ? valueIfNull : value;
   }
+
+  public static long now() {
+    return clock.millis();
+  }
+
+  private static final ZoneId UTC = ZoneId.of("UTC");
+  private static final Map<String, ChronoUnit> dateAddUnits = new HashMap<>();
+
+  static {
+    dateAddUnits.put("years", ChronoUnit.YEARS); // Instant.plus does not support years
+    dateAddUnits.put("months", ChronoUnit.MONTHS); // Instant.plus does not support months
+    dateAddUnits.put("days", ChronoUnit.DAYS);
+    dateAddUnits.put("hours", ChronoUnit.HOURS);
+    dateAddUnits.put("minutes", ChronoUnit.MINUTES);
+    dateAddUnits.put("seconds", ChronoUnit.SECONDS);
+    dateAddUnits.put("millis", ChronoUnit.MILLIS);
+  }
+
+  public static long dateadd(Object input, long delta, String unit) {
+    if (input instanceof String) {
+      return dateadd((String) input, delta, unit);
+    } else if (input instanceof Long) {
+      return dateadd((long) input, delta, unit);
+    }
+
+    throw new IllegalArgumentException(
+        "Invalid input type: "
+            + input.getClass().getSimpleName()
+            + ". Should either be an ISO datetime string like '2007-12-01T12:30:00' or epoch millis");
+  }
+
+  private static long dateadd(String isoDateTime, long delta, String unit) {
+    Instant instant = Instant.parse(isoDateTime);
+    return dateadd(LocalDateTime.ofInstant(instant, UTC), delta, unit);
+  }
+
+  private static long dateadd(long epochMillis, long delta, String unit) {
+    Instant instant = Instant.ofEpochMilli(epochMillis);
+    LocalDateTime localDateTime = LocalDateTime.ofInstant(instant, UTC);
+    return dateadd(localDateTime, delta, unit);
+  }
+
+  private static long dateadd(LocalDateTime localDateTime, long delta, String unit) {
+    ChronoUnit chronoUnit = dateAddUnits.get(unit);
```

Comment:
We could use ChronoUnit.valueOf() and remove the dateAddUnits map ?
```java
  public static long dateadd(Object input, long delta, String unit) {
    if (input instanceof String) {
      return dateadd((String) input, delta, unit);
    } else if (input instanceof Long) {
      return dateadd((long) input, delta, unit);
    }

    throw new IllegalArgumentException(
        "Invalid input type: "
            + input.getClass().getSimpleName()
            + ". Should either be an RFC3339 datetime string like '2007-12-01T12:30:00Z' or epoch millis");
  }

  private static long dateadd(String isoDateTime, long delta, String unit) {
    OffsetDateTime offsetDateTime = OffsetDateTime.parse(isoDateTime);
    return dateadd(offsetDateTime, delta, unit);
  }

  private static long dateadd(long epochMillis, long delta, String unit) {
    Instant instant = Instant.ofEpochMilli(epochMillis);
    OffsetDateTime offsetDateTime = OffsetDateTime.ofInstant(instant, UTC);
    return dateadd(offsetDateTime, delta, unit);
  }

  private static long dateadd(OffsetDateTime offsetDateTime, long delta, String unit) {
    try {
      ChronoUnit chronoUnit = ChronoUnit.valueOf(unit.toUpperCase());
      return offsetDateTime.plus(delta, chronoUnit).toInstant().toEpochMilli();
    } catch (IllegalArgumentException e) {
      throw new IllegalArgumentException(
          "Invalid unit: "
              + unit
              + ". Should be one of "
              + Arrays.stream(ChronoUnit.values())
                  .map(s -> s.name().toLowerCase())
                  .collect(Collectors.toList()),
          e);
    }
  }
```

Reply:
The unit does not map one to one with the Chrono units (it is a subset):
`[nanos, micros, millis, seconds, minutes, hours, half_days, days, weeks, months, years, decades, centuries, millennia, eras, forever]` vs `years, months, days,hours,minutes,millis` - also just preder to have them decoupled in case we use different names (day vs days)

Reply:
You're right. It's better to keep it decoupled.
Maybe use a switch instead of a Map ? (no strong opinion here)
If we keep the Map, we should initialize it with Map.of() (min JDK is 11) instead of a static initializer.

Reply:
+1. Switch fits better, I wanted to cache them for some reason but there is no need to.